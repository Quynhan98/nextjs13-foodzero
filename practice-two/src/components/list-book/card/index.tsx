import { memo, useCallback } from 'react';
import { Button } from '@components/common/button';
import './card.css';

export type CardProp = {
  title: string;
  author: string;
  price: number;
  desc: string;
  image: string;
  id: number;
  onDelete: (id: number) => void;
  onEdit: () => void;
};

const Card = (props: CardProp): JSX.Element => {
  const { title, author, price, desc, image, id, onDelete, onEdit } = props;

  const handleDelete = useCallback((): void => {
    onDelete(id);
  }, [id, onDelete]);

  return (
    <div className="card">
      <div className="card-body">
        <img src={image} alt={`image ${title}`} className="card-image" />
        <h3 className="card-title">{title}</h3>
        <span className="card-author">{author}</span>
        <span className="card-price" data-testid="price">
          Price: {price} USD
        </span>
        <p className="card-desc">{desc}</p>
        <div className="button-group">
          <Button
            data-testid="btnEdit"
            handleClick={onEdit}
            color="btn-secondary"
            typeButton="button"
            size="btn-small"
          >
            Edit
          </Button>
          <Button
            handleClick={handleDelete}
            color="btn-warning"
            typeButton="button"
            size="btn-small"
          >
            Delete
          </Button>
        </div>
      </div>
    </div>
  );
};

export default memo(Card);
