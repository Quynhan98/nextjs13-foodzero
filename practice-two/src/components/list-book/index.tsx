import { Book } from '@interface/book';
import Card from './card';
import './list-book.css';

export type ListBookProp = {
  books: Book[];
  onClickButtonEdit: (item: Book) => void;
  handleRemove: (id: number) => void;
};

export const Books = (props: ListBookProp): JSX.Element => {
  const { books, handleRemove, onClickButtonEdit } = props;

  return (
    <div className="book-list">
      <div className="card-group">
        {books.map((item) => (
          <Card
            onEdit={() => onClickButtonEdit(item)}
            onDelete={handleRemove}
            key={item.id}
            {...item}
          />
        ))}
      </div>
    </div>
  );
};
