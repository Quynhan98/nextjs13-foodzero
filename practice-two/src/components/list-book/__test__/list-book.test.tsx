import { fireEvent, render, screen } from '@testing-library/react';
import { Books, ListBookProp } from '@components/list-book';

describe('Test ListBook Component', () => {
  const books = [
    {
      title: 'Book 1',
      author: 'Author 1',
      price: 1,
      desc: 'Description 1',
      image: 'base_img',
      id: 0,
    },
    {
      title: 'Book 2',
      author: 'Author 2',
      price: 1,
      desc: 'Description 2',
      image: 'base_img',
      id: 1,
    },
  ];
  const initialBooks: ListBookProp = {
    books: books,
    onClickButtonEdit: jest.fn(),
    handleRemove: jest.fn(),
  };

  test('Component ListBook matchers DOM Snapshot', () => {
    const component = render(<Books {...initialBooks} />);

    expect(component).toMatchSnapshot();
  });

  test('Component ListBook should call onEdit function when user click the button', () => {
    render(<Books {...initialBooks} />);

    fireEvent.click(screen.getAllByText('Edit')[0]);
    expect(initialBooks.onClickButtonEdit).toHaveBeenCalled();
  });
});
