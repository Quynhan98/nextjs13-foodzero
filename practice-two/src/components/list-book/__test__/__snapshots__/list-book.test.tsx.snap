// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ListBook Component Component ListBook matchers DOM Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="book-list"
      >
        <div
          class="card-group"
        >
          <div
            class="card"
          >
            <div
              class="card-body"
            >
              <img
                alt="image Book 1"
                class="card-image"
                src="base_img"
              />
              <h3
                class="card-title"
              >
                Book 1
              </h3>
              <span
                class="card-author"
              >
                Author 1
              </span>
              <span
                class="card-price"
                data-testid="price"
              >
                Price: 
                1
                 USD
              </span>
              <p
                class="card-desc"
              >
                Description 1
              </p>
              <div
                class="button-group"
              >
                <button
                  class="button btn-secondary btn-small"
                  type="button"
                >
                  Edit
                </button>
                <button
                  class="button btn-warning btn-small"
                  type="button"
                >
                  Delete
                </button>
              </div>
            </div>
          </div>
          <div
            class="card"
          >
            <div
              class="card-body"
            >
              <img
                alt="image Book 2"
                class="card-image"
                src="base_img"
              />
              <h3
                class="card-title"
              >
                Book 2
              </h3>
              <span
                class="card-author"
              >
                Author 2
              </span>
              <span
                class="card-price"
                data-testid="price"
              >
                Price: 
                1
                 USD
              </span>
              <p
                class="card-desc"
              >
                Description 2
              </p>
              <div
                class="button-group"
              >
                <button
                  class="button btn-secondary btn-small"
                  type="button"
                >
                  Edit
                </button>
                <button
                  class="button btn-warning btn-small"
                  type="button"
                >
                  Delete
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="book-list"
    >
      <div
        class="card-group"
      >
        <div
          class="card"
        >
          <div
            class="card-body"
          >
            <img
              alt="image Book 1"
              class="card-image"
              src="base_img"
            />
            <h3
              class="card-title"
            >
              Book 1
            </h3>
            <span
              class="card-author"
            >
              Author 1
            </span>
            <span
              class="card-price"
              data-testid="price"
            >
              Price: 
              1
               USD
            </span>
            <p
              class="card-desc"
            >
              Description 1
            </p>
            <div
              class="button-group"
            >
              <button
                class="button btn-secondary btn-small"
                type="button"
              >
                Edit
              </button>
              <button
                class="button btn-warning btn-small"
                type="button"
              >
                Delete
              </button>
            </div>
          </div>
        </div>
        <div
          class="card"
        >
          <div
            class="card-body"
          >
            <img
              alt="image Book 2"
              class="card-image"
              src="base_img"
            />
            <h3
              class="card-title"
            >
              Book 2
            </h3>
            <span
              class="card-author"
            >
              Author 2
            </span>
            <span
              class="card-price"
              data-testid="price"
            >
              Price: 
              1
               USD
            </span>
            <p
              class="card-desc"
            >
              Description 2
            </p>
            <div
              class="button-group"
            >
              <button
                class="button btn-secondary btn-small"
                type="button"
              >
                Edit
              </button>
              <button
                class="button btn-warning btn-small"
                type="button"
              >
                Delete
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
